### PR TITLE
[WIP] wallet: ensure wallet files are not reused across chains

### PR DESCRIPTION
**Pull request by rodentrabies**
This implements a proposal in #12805.

I follow the suggestion by @ryanofsky and compare a genesis block hash, extracted from wallet's best block locator, to an `chainActive`'s genesis block hash. If they don't match, the error is reported.

### CODE REVIEW

**Comment by gmaxwell** [comment]
Genesis block isn't sufficient, a lot of altcoins share genesis blocks. :(  Also there should be a way to override, since moving wallet files between forks is a common way of collecting fork coins-- obviously not something you'd want to happen accidentally but fine as an unsupported "you get to keep both pieces" option.

**Comment by mrwhythat** [comment]
@gmaxwell, thanks for the suggestion. Just added an override option `-allowreusewallets` (maybe should be renamed to something clearer). Still working on sufficiency issue. First of all, since `CChain::GetLocator` builds a `vHave` sequence deterministically, maybe the whole vectors can be compared, but a lot of forked altcoins share those blocks as well, so I'm looking for an alternative.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -4095,13 +4095,28 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
 
     LOCK(cs_main);
 
-    CBlockIndex *pindexRescan = chainActive.Genesis();
-    if (!gArgs.GetBoolArg("-rescan", false))
+    CBlockIndex* activeChainGenesis = chainActive.Genesis();
+    WalletBatch batch(*walletInstance->database);
```
[the remark this answers is not in the document]
Comment (mrwhythat) [reply src/wallet/wallet.cpp]: It's done on purpose in order to reuse the value, computed by it (`bestBlockRead`) both here and in `-rescan` check.

**Comment by mrwhythat** [comment]
Implemented the new check via `FindForkInGlobalIndex`. This solution looks at all blocks known to the wallet and not only at a genesis. It may misbehave in case of temporary forks (reorgs), but there seems to be no simple way around that.

This can be tested with the following function test (here two disconnected nodes in regtest mode generate two different chains with the same genesis, so can be used to model fork coins with shared genesis):
```python
        # ...
        self.log.info("Generating initial distinct blockchains")
        self.nodes[0].generate(5)
        self.nodes[1].generate(5)

        self.log.info("Creating wallets")
        w0_path = os.path.join(self.nodes[0].datadir, 'w0')
        w1_path = os.path.join(self.nodes[1].datadir, 'w1')
        self.nodes[0].createwallet(w0_path)
        self.nodes[1].createwallet(w1_path)

        self.log.info("Restarting nodes")  # to close wallets
        self.stop_node(0)
        self.stop_node(1)
        self.start_node(0)
        self.start_node(1)

        self.log.info("Loading wallet")
        print(self.nodes[0].loadwallet(w1_path))  # this should fail with wallet reuse error
```
Not sure if this needs a separate test though, so any suggestions on which one should have it appended?

**Comment by MeshCollider** [comment]
It doesn't seem like a good idea to prevent a wallet from opening in the case of even a 1 block reorg IMO. Maybe checking 6 blocks back in the chain or something would be better?

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -4095,6 +4095,24 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
 
     LOCK(cs_main);
 
+    // Unless allowed, ensure wallet files are not reused across chains:
+    if (!gArgs.GetBoolArg("-allowreusewallets", false))
```
Comment (MeshCollider) [inline src/wallet/wallet.cpp]: `DEFAULT_ALLOWREUSEWALLETS `

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -4095,6 +4095,24 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
...
+        WalletBatch batch(*walletInstance->database);
+        CBlockLocator locator;
+        if (batch.ReadBestBlock(locator))
+        {
+            // Wallet is assumed to be from another chain, if its best known
+            // block is not in active chain:
+            CBlockIndex *forkIndex = FindForkInGlobalIndex(chainActive, locator);
+            if (locator.vHave.size() > 0 && forkIndex && forkIndex->GetBlockHash() != locator.vHave.front())
+            {
+                InitError(_("Wallet files should not be reused across chains"));
```
Comment (MeshCollider) [inline src/wallet/wallet.cpp]: Maybe `walletInstance->WalletLogPrintf` the hashes or something too

File: src/wallet/init.cpp [hunk 4]
```diff
@@ -38,6 +38,7 @@ const WalletInitInterface& g_wallet_init_interface = WalletInit();
 void WalletInit::AddWalletOptions() const
 {
     gArgs.AddArg("-addresstype", strprintf("What type of addresses to use (\"legacy\", \"p2sh-segwit\", or \"bech32\", default: \"%s\")", FormatOutputType(DEFAULT_ADDRESS_TYPE)), false, OptionsCategory::WALLET);
+    gArgs.AddArg("-allowreusewallets", strprintf("Allow reusing wallet files across chains (default: %u)", DEFAULT_ALLOWREUSEWALLETS), false, OptionsCategory::WALLET);
```
Comment (luke-jr) [inline src/wallet/init.cpp]: I'm not sure this should really be an option?

Comment (luke-jr) [inline src/wallet/init.cpp]: If there is an option, the name `-walletcrosschain` might be better.

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -4126,6 +4126,24 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
...
 
+    // Unless allowed, ensure wallet files are not reused across chains:
+    if (!gArgs.GetBoolArg("-allowreusewallets", DEFAULT_ALLOWREUSEWALLETS))
+    {
+        WalletBatch batch(*walletInstance->database);
+        CBlockLocator locator;
+        if (batch.ReadBestBlock(locator))
+        {
+            // Wallet is assumed to be from another chain, if its best known
+            // block is not in active chain:
```
Comment (luke-jr) [inline src/wallet/wallet.cpp]: This comment's described behaviour is definitely wrong: the best known block could be a stale block!

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (mrwhythat) [reply src/wallet/wallet.cpp]: Indeed, but this is going away after the @MeshCollider's comments are addressed. I didn't have much time lately, will do it sometime the next two weeks.

**Comment by ryanofsky** [comment]
Not sure, but it may be more important to add this feature with the new "Open Wallet -> Other..." GUI feature in https://github.com/bitcoin/bitcoin/pull/15204 to prevent mistakenly loading incompatible wallets.

**Comment by mrwhythat** [comment]
Looks like I completely forgot about this one. Updated and tested the old branch yesterday - https://github.com/mrwhythat/bitcoin/tree/wallet-file-reuse-prevention. Can it be reopened or should I just open another PR?

**Comment by fanquake** [comment]
I'm unable to re-open this PR, so creating a new one should be ok.

**Pull request closed without merging**